python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py

```diff
@@ -195,26 +214,95 @@ def visit_Expr(self, node):
                     return
                 else:
                     self._visit_Call(child_node)
+        return node
+
+    def visit_Attribute(self, node):
+        if self.used_by_paddle_api(node):
+            if self.is_tensor_shape(node):
+                return create_api_shape_node(node)
+        return node
 
+    def visit_Name(self, node):
+        if node.id in self.name_to_tensor_shape:
+            if self.used_by_paddle_api(node):
+                tensor_shape_node = self.name_to_tensor_shape[node.id]
+                if isinstance(tensor_shape_node, gast.Attribute):
+                    return create_api_shape_node(tensor_shape_node)
+                elif isinstance(tensor_shape_node, gast.Subscript):
+                    result_node = copy.deepcopy(tensor_shape_node)
+                    result_node.value = create_api_shape_node(
+                        tensor_shape_node.value)
+                    return result_node
         return node
 
     def _visit_Call(self, node):
         assert isinstance(node, gast.Call)
-
         # Replace API `to_variable` with `fluid.layers.assign`
         if is_to_variable(node):
             self._update_feed_dict(node)
             node = to_assign_node(node)
             return node
 
+        if is_paddle_api(node):
+            # Visit gast.Attribute and gast.Name to replace tensor.shape if necessary
+            self.generic_visit(node)
+
         func_name = astor.to_source(gast.gast_to_ast(node.func))
+
         if self._is_dygraph_forward(func_name):
             class_node = self._get_class_node(func_name)
             static_node = to_static_ast(node, class_node)
             return static_node
         else:
             return node
 
+    def is_tensor_shape(self, node):
+        """
+        Return True if node is like `x.shape` and x is Tensor, return False otherwise.
+        """
+        assert isinstance(node, gast.Attribute)
+        if node.attr != 'shape':
+            return False
+
+        try:
+            value_id = node.value.id
+        except AttributeError:
+            return False
+
+        if value_id in self.name_to_tensor_shape:
+            return True
+
+        # TODO: `value_id` may be not in scope_var_type_dict if `value_id` is the arg of decorated function
+        # Need a better way to confirm whether `value_id` is a Tensor.
+        try:
+            var_type_set = self.scope_var_type_dict[value_id]
+        except KeyError:
+            return False
+
+        if NodeVarType.NUMPY_NDARRAY in var_type_set:
+            return False
+        if NodeVarType.TENSOR not in var_type_set and NodeVarType.PADDLE_RETURN_TYPES not in var_type_set:
+            return False
+
+        return True
+
+    def used_by_paddle_api(self, node):
```

Comment:
If `used_by_paddle_api ` is private function, `_used_by_paddle_api ` will be better.

Reply:
Thanks.